### PR TITLE
feat(linter): add oxc/no-redundant-constructor-init

### LINKED ISSUE #9267: linter: new rule - redundant initialization in class constructor
```ts
class Foo {
  constructor(public name: unknown) {
    this.name = name;
  }
}
```

generates

```js
class Foo {
    constructor(name) {
        this.name = name;
        this.name = name;
    }
}
```

which is redundant.

Inspired by https://github.com/rolldown/rolldown/pull/3651

### PR DESCRIPTION
**Pull request by BenShelton**
closes #9267

I'm unsure whether the rule naming is correct, I believe this is a rule that doesn't exist elsewhere so came up with the name `no_redundant_constructor_init`, if that needs to be changed or if the rule should be moved elsewhere then happy to do so.

### CODE REVIEW

File: crates/oxc_linter/src/rules.rs [hunk 1]
```diff
@@ -630,6 +631,7 @@ oxc_macros::declare_all_lint_rules! {
     eslint::no_proto,
     eslint::no_prototype_builtins,
     eslint::no_redeclare,
+    eslint::no_redundant_constructor_init,
```
Comment (shulaoda) [inline crates/oxc_linter/src/rules.rs:634]: I think it should be classified as plugin `oxc` instead of `eslint`.

In addition, we should add this test case to fail:
```js
class Foo {
  constructor(other: unknown, name: unknown) {
    this.name = name;
    this.other = other;
    this.name = name;
  }
}
```

**Comment by Boshen** [comment]
Thank you so much for working on this! Here are some results from running this rule in https://github.com/oxc-project/oxlint-ecosystem-ci

Kibana 

```
...

  × eslint(no-redundant-constructor-init): Explicit initialization of public members is redundant
    ╭─[src/plugins/data/common/search/aggs/utils/date_interval_utils/invalid_es_calendar_interval_error.ts:29:5]
 28 │     this.unit = unit;
 29 │     this.type = type;
    ·     ─────────────────
 30 │
    ╰────
  help: Remove the explicit initialization

  × eslint(no-redundant-constructor-init): Explicit initialization of public members is redundant
    ╭─[packages/kbn-apm-synthtrace-client/src/lib/entity.ts:15:5]
 14 │   constructor(public readonly fields: TFields) {
 15 │     this.fields = fields;
    ·     ─────────────────────
 16 │   }
    ╰────
  help: Remove the explicit initialization

Found 0 warnings and 9 errors.
```

vscode:

```
...

  × eslint(no-redundant-constructor-init): Explicit initialization of public members is redundant
    ╭─[src/vs/editor/common/services/modelService.ts:41:3]
 40 │     ) {
 41 │         this.model = model;
    ·         ───────────────────
 42 │         this._modelEventListeners.add(model.onWillDispose(() => onWillDispose(model)));
    ╰────
  help: Remove the explicit initialization

  × eslint(no-redundant-constructor-init): Explicit initialization of public members is redundant
    ╭─[src/vs/workbench/contrib/testing/browser/testResultsView/testResultsSubject.ts:53:3]
 52 │         const messages = test.tasks[taskIndex].messages;
 53 │         this.messageIndex = messageIndex;
    ·         ─────────────────────────────────
 54 │
    ╰────
  help: Remove the explicit initialization

Found 0 warnings and 5 errors.
```

File: crates/oxc_linter/src/rules/eslint/no_redundant_constructor_init.rs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// class Foo {
+    ///   constructor(public name: unknown) {}
+    /// }
+    /// ```
+    NoRedundantConstructorInit,
+    eslint,
+    style,
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/eslint/no_redundant_constructor_init.rs:49]: This is definitely a `correctness` rule, because the offending code is useless, or a mistake.

File: crates/oxc_linter/src/rules.rs (hunk 1 above)
Comment (BenShelton) [reply crates/oxc_linter/src/rules.rs:634]: That makes sense to move to `oxc` instead of `eslint`, will do that.

For the extra test case I feel like the main goal of this rule was from `public` arguments not understood as being automatically initialized and it wouldn't be obvious unless you look at the generated output that the assignment is duplicated. This test case is just a mistake of the user. If we do decide to go down enforcing that too we'd have to start catching edge cases such as reassigning the argument:

```ts
class Foo {
  constructor(other: unknown, name: unknown) {
    this.name = name;
    this.other = other;
    if (someCase) {
      name = 'new';
      this.name = name;
    }
  }
}
```

Given that, is this still a use case we want to enforce with this rule?

File: crates/oxc_linter/src/rules/eslint/no_redundant_constructor_init.rs (hunk 2 above)
Comment (BenShelton) [reply crates/oxc_linter/src/rules/eslint/no_redundant_constructor_init.rs:49]: Thanks for clarifying

**Comment by BenShelton** [comment]
Tests were failing because a load of snapshots had a line similar to the following where the number of rules increased by 1:

```diff
- Finished in <variable>ms on 1 file with 99 rules using 1 threads.
+ Finished in <variable>ms on 1 file with 100 rules using 1 threads.
```

I've updated and committed all those snapshots, hopefully that is correct.

**Review by shulaoda — APPROVED** [review]

File: crates/oxc_linter/src/rules/oxc/no_redundant_constructor_init.rs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                |param| if param.is_public() { param.pattern.get_identifier_name() } else { None },
+            )
+            .collect::<Vec<_>>();
+        if public_members.is_empty() {
+            return;
+        }
+        let Some(body) = method.value.body.as_ref() else {
+            return;
+        };
+        body.statements.iter().for_each(|stmt| {
```
Comment (camc314) [inline crates/oxc_linter/src/rules/oxc/no_redundant_constructor_init.rs:76]: probably not for now, but i wonder whether it'd be better to grab the public members references and check that way. WIth this approach, if you had it assigned in a block, it wouldn't report an err

**Review by camc314 — APPROVED** [review]
💪

File: crates/oxc_linter/src/rules/oxc/no_redundant_constructor_init.rs (hunk 3 above)
Comment (BenShelton) [reply crates/oxc_linter/src/rules/oxc/no_redundant_constructor_init.rs:76]: Nice, didn’t know that was available. Will definitely improve in a follow up.
